File: iosApp/Views/Onboarding/Microscopist/InitialsStep.swift
import SwiftUI

/// Phase 2 step 2 (spec §10): optional 2-character initials or skip.
/// Initials appear next to the microscopist's identifier in audit
/// reports so reviewers can recognise who recorded an override.
struct InitialsStep: View {

    @Environment(OnboardingState.self) private var onboarding

    @State private var initials: String = ""
    @State private var errorMessage: String?

    private var trimmed: String {
        initials.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    var body: some View {
        WizardStepContainer(
            title: "Your initials",
            subtitle: "Up to two characters. These appear next to your overrides in the audit log. Optional — you can skip this.",
            stepIndicator: "Step 2 of 4",
            primary: .init(title: "Continue", isEnabled: trimmed.count <= 2) {
                save(initials: trimmed.isEmpty ? nil : trimmed)
            },
            secondary: .init(title: "Skip") {
                save(initials: nil)
            }
        ) {
            VStack(alignment: .leading, spacing: 8) {
                Text("Initials")
                    .font(.headline)
                TextField("e.g. JM", text: $initials)
                    .textFieldStyle(.roundedBorder)
                    .autocorrectionDisabled(true)
                    .textInputAutocapitalization(.characters)
                    .submitLabel(.done)
                    .onChange(of: initials) { _, newValue in
                        // Clamp to 2 characters as the user types — the
                        // spec column is exactly 2 chars wide and longer
                        // strings would mis-render in the audit list.
                        if newValue.count > 2 {
                            initials = String(newValue.prefix(2))
                        }
                    }
            }
        }
        .alert(
            "Unable to save initials",
            isPresented: Binding(
                get: { errorMessage != nil },
                set: { if !$0 { errorMessage = nil } }
            )
        ) {
            Button("OK", role: .cancel) { errorMessage = nil }
        } message: {
            Text(errorMessage ?? "")
        }
    }

    private func save(initials: String?) {
        do {
            try onboarding.setMicroscopistInitials(initials)
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}

File: androidApp/src/main/kotlin/com/malaria/android/ui/onboarding/microscopist/InitialsStep.kt
package com.malaria.android.ui.onboarding.microscopist

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.unit.dp
import com.malaria.android.ui.locals.LocalOnboardingState
import com.malaria.android.ui.onboarding.components.WizardStepContainer
import kotlinx.coroutines.launch

/**
 * Phase 2 step 2 (spec §10): optional 2-character initials or skip.
 * Initials appear next to the microscopist's identifier in audit reports so
 * reviewers can recognise who recorded an override.
 *
 * Mirrors `iosApp/Views/Onboarding/Microscopist/InitialsStep.swift`.
 */
@Composable
fun InitialsStep() {
    val onboarding = LocalOnboardingState.current
    val scope = rememberCoroutineScope()

    var initials by remember { mutableStateOf("") }
    val trimmed = initials.trim()

    WizardStepContainer(
        title = "Your initials",
        subtitle = "Up to two characters. These appear next to your overrides in the audit log. Optional — you can skip this.",
        stepIndicator = "Step 2 of 4",
        primaryLabel = "Continue",
        primaryEnabled = trimmed.length <= 2,
        onPrimary = {
            scope.launch {
                onboarding.setMicroscopistInitials(if (trimmed.isEmpty()) null else trimmed)
            }
        },
        secondaryLabel = "Skip",
        onSecondary = {
            scope.launch { onboarding.setMicroscopistInitials(null) }
        },
    ) {
        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Text("Initials", style = MaterialTheme.typography.titleSmall)
            OutlinedTextField(
                value = initials,
                onValueChange = { newValue ->
                    // Clamp to 2 characters as the user types — the spec
                    // column is exactly 2 chars wide.
                    initials = if (newValue.length > 2) newValue.take(2) else newValue
                },
                placeholder = { Text("e.g. JM") },
                singleLine = true,
                keyboardOptions = KeyboardOptions(
                    capitalization = KeyboardCapitalization.Characters,
                    imeAction = ImeAction.Done,
                ),
                modifier = Modifier.fillMaxWidth(),
            )
        }
    }
}
